Provide a specific error for a rule specification missing a plugin prefix

**The version of ESLint you are using.** 

7.18.0

**The problem you want to solve.**

In trying to enable a rule from a new custom plugin in my `.eslintrc.js`, I made the mistake of forgetting to include the plugin prefix in the specifier:

```js
{
    "extends": [
      "plugin:jest-react/recommended"
    ],
    "plugins": [
      "jest-react"
    ],
    "rules": {
        // This should have been:
        // "jest-react/no-mocking-react": "error"
        "no-mocking-react": "error"
    }
  }
```

**Your take on the correct solution to problem.**

It would be nice if a _"did you mean?"_ could be printed at the end of all outputs for each rule that could be matched but for a plugin prefix. Roughly proposing:

```text
/path/to/file1.js
  1:1  error  Definition for rule 'no-mocking-react' was not found (though a jest/  no-mocking-react

Definition for 'no-mocking-react' was not found, but 'jest-react/no-mocking-react' was.
```

**Are you willing to submit a pull request to implement this change?**

It would be my honor and privilege. ❤️ 

I dug a little bit and saw that rules are stored in `slots.lastConfigArray.pluginRules` within `Linter`. If a rule is not found, we could check if (semi-pseudocoding here)...

```js
Array.from(slots.lastConfigArray.pluginRules.entries()).some(ruleEntryKey => ruleEntryKey === `/${id}`)
```

--- a/lib/config/rule-validator.js
+++ b/lib/config/rule-validator.js
@@ -35,16 +35,20 @@ function findRuleDefinition(ruleId, config) {
         pluginName = ruleIdParts.join("/");
     }
 
-    if (config.plugins) {
+    const errorMessageHeader = `Key "rules": Key "${ruleId}"`;
+    let errorMessage = `${errorMessageHeader}: Could not find plugin "${pluginName}".`;
+
+    // if the plugin exists then we need to check if the rule exists
+    if (config.plugins && config.plugins[pluginName]) {
 
         const plugin = config.plugins[pluginName];
 
         // first check for exact rule match
-        if (plugin && plugin.rules && plugin.rules[ruleName]) {
+        if (plugin.rules && plugin.rules[ruleName]) {
             return config.plugins[pluginName].rules[ruleName];
         }
 
-        let errorMessage = `Key "rules": Key "${ruleId}": Could not find "${ruleName}" in plugin "${pluginName}".`;
+        errorMessage = `${errorMessageHeader}: Could not find "${ruleName}" in plugin "${pluginName}".`;
 
         // otherwise, let's see if we can find the rule name elsewhere
         for (const [otherPluginName, otherPlugin] of Object.entries(config.plugins)) {
@@ -54,11 +58,10 @@ function findRuleDefinition(ruleId, config) {
             }
         }
 
-        throw new TypeError(errorMessage);
-
+        // falls through to throw error
     }
 
-    throw new TypeError(`Key "rules": Key "${ruleId}": Could not find plugin "${pluginName}".`);
+    throw new TypeError(errorMessage);
 }
 
 /**

--- a/lib/config/rule-validator.js
+++ b/lib/config/rule-validator.js
@@ -35,18 +35,20 @@ function findRuleDefinition(ruleId, config) {
         pluginName = ruleIdParts.join("/");
     }
 
-    if (config?.plugins) {
+    if (config.plugins) {
+
+        const plugin = config.plugins[pluginName];
 
         // first check for exact rule match
-        if (config.plugins?.[pluginName]?.rules?.[ruleName]) {
+        if (plugin && plugin.rules && plugin.rules[ruleName]) {
             return config.plugins[pluginName].rules[ruleName];
         }
 
         let errorMessage = `Key "rules": Key "${ruleId}": Could not find "${ruleName}" in plugin "${pluginName}".`;
 
         // otherwise, let's see if we can find the rule name elsewhere
-        for (const [otherPluginName, plugin] of Object.entries(config.plugins)) {
-            if (plugin?.rules?.[ruleName]) {
+        for (const [otherPluginName, otherPlugin] of Object.entries(config.plugins)) {
+            if (otherPlugin.rules && otherPlugin.rules[ruleName]) {
                 errorMessage += ` Did you mean "${otherPluginName}/${ruleName}"?`;
                 break;
             }

--- a/lib/config/rule-validator.js
+++ b/lib/config/rule-validator.js
@@ -35,16 +35,28 @@ function findRuleDefinition(ruleId, config) {
         pluginName = ruleIdParts.join("/");
     }
 
-    if (!config.plugins || !config.plugins[pluginName]) {
-        throw new TypeError(`Key "rules": Key "${ruleId}": Could not find plugin "${pluginName}".`);
+    if (config?.plugins) {
+
+        // first check for exact rule match
+        if (config.plugins?.[pluginName]?.rules?.[ruleName]) {
+            return config.plugins[pluginName].rules[ruleName];
+        }
+
+        let errorMessage = `Key "rules": Key "${ruleId}": Could not find "${ruleName}" in plugin "${pluginName}".`;
+
+        // otherwise, let's see if we can find the rule name elsewhere
+        for (const [otherPluginName, plugin] of Object.entries(config.plugins)) {
+            if (plugin?.rules?.[ruleName]) {
+                errorMessage += ` Did you mean "${otherPluginName}/${ruleName}"?`;
+                break;
+            }
+        }
+
+        throw new TypeError(errorMessage);
+
     }
 
-    if (!config.plugins[pluginName].rules || !config.plugins[pluginName].rules[ruleName]) {
-        throw new TypeError(`Key "rules": Key "${ruleId}": Could not find "${ruleName}" in plugin "${pluginName}".`);
-    }
-
-    return config.plugins[pluginName].rules[ruleName];
-
+    throw new TypeError(`Key "rules": Key "${ruleId}": Could not find plugin "${pluginName}".`);
 }
 
 /**

--- a/tests/lib/config/flat-config-array.js
+++ b/tests/lib/config/flat-config-array.js
@@ -56,6 +56,16 @@ const baseConfig = {
                     }
                 }
             }
+        },
+        test1: {
+            rules: {
+                match: {}
+            }
+        },
+        test2: {
+            rules: {
+                nomatch: {}
+            }
         }
     }
 };
@@ -1276,6 +1286,28 @@ describe("FlatConfigArray", () => {
                         }
                     }
                 ], "Key \"rules\": Key \"foo\": Expected severity of \"off\", 0, \"warn\", 1, \"error\", or 2.");
+            });
+
+            it("should error when rule doesn't exist", async () => {
+
+                await assertInvalidConfig([
+                    {
+                        rules: {
+                            foox: [1, "bar"]
+                        }
+                    }
+                ], /Key "rules": Key "foox": Could not find "foox" in plugin "@"./u);
+            });
+
+            it("should error and suggest alternative when rule doesn't exist", async () => {
+
+                await assertInvalidConfig([
+                    {
+                        rules: {
+                            "test2/match": "error"
+                        }
+                    }
+                ], /Key "rules": Key "test2\/match": Could not find "match" in plugin "test2"\. Did you mean "test1\/match"\?/u);
             });
 
             it("should error when rule options don't match schema", async () => {

--- a/tests/lib/config/flat-config-array.js
+++ b/tests/lib/config/flat-config-array.js
@@ -1310,6 +1310,17 @@ describe("FlatConfigArray", () => {
                 ], /Key "rules": Key "test2\/match": Could not find "match" in plugin "test2"\. Did you mean "test1\/match"\?/u);
             });
 
+            it("should error when plugin for rule doesn't exist", async () => {
+
+                await assertInvalidConfig([
+                    {
+                        rules: {
+                            "doesnt-exist/match": "error"
+                        }
+                    }
+                ], /Key "rules": Key "doesnt-exist\/match": Could not find plugin "doesnt-exist"\./u);
+            });
+
             it("should error when rule options don't match schema", async () => {
 
                 await assertInvalidConfig([

